### PR TITLE
feat(flow): flow worker

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

A simple flow worker that runs on a seperate thread and accept requests accross thread in a blocking style

Please explain IN DETAIL what the changes are in this PR and why they are needed:

A flow worker which can create flow, drop flow and trigger flow running

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/flow/src/adapter/worker.rs [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+    let server = InterThreadCallServer {
+        arg_recv,
+        ret_sender: ret_send,
+    };
+    (client, server)
+}
+
+#[derive(Debug)]
+struct InterThreadCallClient {
+    call_id: Arc<Mutex<usize>>,
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:402]: atomic instead?

File: src/flow/src/adapter/worker.rs [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+
+impl InterThreadCallClient {
+    /// call without expecting responses or blocking
+    async fn call_non_blocking(&self, req: Request) {
+        let call_id = {
+            let mut call_id = self.call_id.lock().await;
+            *call_id += 1;
+            *call_id
+        };
+        self.arg_sender.send((call_id, req)).unwrap();
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:415]: Please handle unwrap

File: src/flow/src/adapter/worker.rs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+    }
+    /// call blocking, and return the result
+    async fn call_blocking(&mut self, req: Request) -> Result<Response, Error> {
+        let call_id = {
+            let mut call_id = self.call_id.lock().await;
+            *call_id += 1;
+            *call_id
+        };
+        self.arg_sender.send((call_id, req)).unwrap();
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:424]: ditto

File: src/flow/src/adapter/worker.rs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+        // TODO(discord9): better inter thread call impl
+        let (ret_call_id, ret) = self.ret_recv.recv().await.unwrap();
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:426]: dito

File: src/flow/src/adapter/worker.rs [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+        if ret_call_id != call_id {
+            return InternalSnafu {
+                reason: "call id mismatch, worker/worker handler should be in sync",
+            }
+            .fail()
+            .with_context(|_| EvalSnafu {});
+        }
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:433]: Can replace with `ensure!`

File: src/flow/src/adapter/worker.rs [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+        self.arg_recv.recv().await
+    }
+
+    pub fn blocking_recv(&mut self) -> Option<(usize, Request)> {
+        self.arg_recv.blocking_recv()
+    }
+
+    /// Send response back to the client
+    pub fn resp(&self, call_id: usize, resp: Response) {
+        self.ret_sender.send((call_id, resp)).unwrap();
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:455]: handle unwrap unless it's safe

File: src/flow/src/adapter/worker.rs [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+
+    pub async fn contains_flow(&self, task_id: FlowId) -> Result<bool, Error> {
+        let req = Request::ContainTask { task_id };
+        let ret = self
+            .itc_client
+            .lock()
+            .await
+            .call_blocking(req)
+            .await
+            .unwrap();
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:187]: handle unwrap unless it's safe

File: src/flow/src/adapter/worker.rs [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+        sink_sender: mpsc::UnboundedSender<DiffRow>,
+        source_ids: &[GlobalId],
+        src_recvs: Vec<broadcast::Receiver<DiffRow>>,
+        // TODO(discord9): set expire duration for all arrangement and compare to sys timestamp instead
+        expire_when: Option<repr::Duration>,
+        create_if_not_exist: bool,
+        err_collector: ErrCollector,
+    ) -> Result<Option<FlowId>, Error> {
+        let _ = expire_when;
+        if create_if_not_exist {
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:238]: Should an error be reported when `create_if_not_exist == false` and `task_id` already exist?

File: src/flow/src/adapter/worker.rs (hunk 1 above)
Comment (fengjiachun) [reply src/flow/src/adapter/worker.rs:402]: +1

File: src/flow/src/adapter/worker.rs [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+                    "Flow Node/Worker itc failed, expect Response::Create, found {ret:?}"
+                ),
+            }
+            .fail()
+            .with_context(|_| EvalSnafu {})
+        }
+    }
+
+    /// remove task, return task id
+    pub async fn remove_flow(&self, task_id: FlowId) -> Result<bool, Error> {
```
Comment (fengjiachun) [inline src/flow/src/adapter/worker.rs:154]:
```suggestion
    pub async fn remove_flow(&self, flow_id: FlowId) -> Result<bool, Error> {
```

File: src/flow/src/adapter/worker.rs [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+        task_id: FlowId,
+        plan: TypedPlan,
+        sink_id: GlobalId,
+        sink_sender: mpsc::UnboundedSender<DiffRow>,
+        source_ids: &[GlobalId],
+        src_recvs: Vec<broadcast::Receiver<DiffRow>>,
+        expire_when: Option<repr::Duration>,
+        create_if_not_exist: bool,
+        err_collector: ErrCollector,
+    ) -> Result<Option<FlowId>, Error> {
```
Comment (fengjiachun) [inline src/flow/src/adapter/worker.rs:123]: Can we just use the `Request::Create` as the input args?

File: src/flow/src/adapter/worker.rs [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+            task_create_result
+        } else {
+            InternalSnafu {
+                reason: format!(
+                    "Flow Node/Worker itc failed, expect Response::Create, found {ret:?}"
+                ),
+            }
+            .fail()
+            .with_context(|_| EvalSnafu {})
+        }
```
Comment (zhongzc) [inline src/flow/src/adapter/worker.rs:150]: Introducing `enum-as-inner` can keep thing compact:

```rust
#[derive(Debug, enum_as_inner::EnumAsInner)]
enum Response {
```

```suggestion
        ret.into_create()
            .map_err(|ret| {
                InternalSnafu {
                    reason: format!(
                        "Flow Node/Worker itc failed, expect Response::Create, found {ret:?}"
                    ),
                }
                .build()
            })
            .context(EvalSnafu)?
```

File: src/flow/src/adapter.rs [hunk 12]
```diff
@@ -19,3 +19,12 @@ pub(crate) mod error;
...
 pub(crate) use node_context::FlownodeContext;
+
+mod worker;
+
+pub const PER_REQ_MAX_ROW_CNT: usize = 8192;
+
+// TODO: refactor common types for flow to a separate module
+/// FlowId is a unique identifier for a flow task
+pub type FlowId = u64;
+pub type TableName = [String; 3];
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/flow/src/adapter.rs:30]: > Where have I seen them(FlowId, TableName) before, is it a repetition here?

Thanks for noticing, I must missed it during cherry pick

File: src/flow/src/adapter/worker.rs [hunk 13]
```diff
@@ -109,50 +109,35 @@ impl WorkerHandle {
...
-            plan,
-            sink_id,
-            sink_sender,
-            source_ids: source_ids.to_vec(),
-            src_recvs,
-            expire_when,
-            create_if_not_exist,
-            err_collector,
-        };
+    pub async fn create_flow(&self, create_reqs: Request) -> Result<Option<FlowId>, Error> {
```
Comment (fengjiachun) [inline src/flow/src/adapter/worker.rs:112]: What about: `create_reqs: Request::Create` ?

Comment (discord9) [reply src/flow/src/adapter/worker.rs:112]: > What about: `create_req: Request::Create` ?

seems not possible: https://users.rust-lang.org/t/passing-enum-variant-as-argument/94422/1

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**